### PR TITLE
Fix double-click in data-grid if lazy-loading (T1208575)

### PR DESCRIPTION
**Pull request by GoodDayForSurf**

### CODE REVIEW

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/mask.tests.js [hunk 1]
```diff
@@ -2087,13 +2087,15 @@ QUnit.module('format: caret boundaries', moduleConfig, () => {
...
-        }
 
+        this.keyboard.caret(3);
+        this.input.trigger('dxclick');
+        this.clock.tick(CARET_TIMEOUT_DURATION);
+
+        assert.timeout(100);
         assert.deepEqual(this.keyboard.caret(), { start: 1, end: 1 }, 'caret is on integer part end');
+
+        this.clock.restore();
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/mask.tests.js]: It's restored in `afterEach`

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/mask.tests.js [hunk 2]
```diff
@@ -2087,13 +2087,15 @@ QUnit.module('format: caret boundaries', moduleConfig, () => {
...
-            this.keyboard.caret(3);
-            this.input.trigger('dxclick');
-            this.clock.tick(CARET_TIMEOUT_DURATION);
-        }
 
+        this.keyboard.caret(3);
+        this.input.trigger('dxclick');
+        this.clock.tick(CARET_TIMEOUT_DURATION);
+
+        assert.timeout(100);
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/mask.tests.js]: We usually use sinon for timers as a more strict way.
Why do we need this magic timeout?

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/mask.tests.js [hunk 3]
```diff
@@ -2087,13 +2087,15 @@ QUnit.module('format: caret boundaries', moduleConfig, () => {
 
         this.input.focus();
         this.clock.tick(CARET_TIMEOUT_DURATION);
-        for(let i = 0; i < 2; ++i) {
-            this.keyboard.caret(3);
-            this.input.trigger('dxclick');
-            this.clock.tick(CARET_TIMEOUT_DURATION);
-        }
 
+        this.keyboard.caret(3);
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/mask.tests.js]: Tbh i don't know why loop was used here but I also don't see a reason to change it.
Did u check that this test didn't become evergreen after your change?

File: packages/devextreme/js/events/double_click.js [hunk 4]
```diff
@@ -57,9 +57,11 @@ const DblClick = Class.inherit({
 
     remove: function() {
         this._handlerCount--;
+
         if(this._handlerCount <= 0) {
             this._forgetLastClick();
             eventsEngine.off(domAdapter.getDocument(), NAMESPACED_CLICK_EVENT);
+            this._handlerCount = 0;
```
Comment (ksercs) [inline packages/devextreme/js/events/double_click.js]: I'm not going to review this change but I believe there should be a test for this

File: packages/devextreme/js/ui/number_box/number_box.mask.js [hunk 5]
```diff
@@ -579,7 +580,7 @@ const NumberBoxMask = NumberBoxBase.inherit({
             }
         }.bind(this));
 
-        eventsEngine.on($input, 'dxdblclick', function() {
+        eventsEngine.on($input, dxDblClickEvent, function() {
```
Comment (ksercs) [inline packages/devextreme/js/ui/number_box/number_box.mask.js:583]: so, u wanna say it has never worked before?

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/mask.tests.js (hunk 3 above)
Comment (GoodDayForSurf) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/mask.tests.js]: it seems, this test checks returning caret after click. So after my fix,two dxclick fire dxdblclick therefore loop is not correct test here

File: packages/devextreme/js/__internal/events/dblclick_impl.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
+import Class from '@js/core/class';
```
Comment (ksercs) [inline packages/devextreme/js/__internal/events/dblclick_impl.ts]: Why did u add "_impl" as a suffix in the file name?

File: packages/devextreme/testing/tests/DevExpress.ui.events/dblclick.tests.js [hunk 7]
```diff
@@ -34,6 +35,21 @@ QUnit.test('dxdblclick should bubble up', function(assert) {
     $('#element').trigger(dblclickEvent.name);
 });
 
+QUnit.test('dxdblclick should be works correctly even if dblClick.remove() calls many times (T1208575)', function(assert) {
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.events/dblclick.tests.js]: maybe this way?
```suggestion
QUnit.test('dxdblclick should not be handled on a usual dxclick even if dblClick.remove() was called before (T1208575)', function(assert) {
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/mask.tests.js (hunk 3 above)
Comment (ksercs) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/mask.tests.js]: And still, did u check that this test is not evergreen now? It was added during a bug fix

File: packages/devextreme/js/__internal/events/dblclick_impl.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+import { name as clickEventName } from '@js/events/click';
+import eventsEngine from '@js/events/core/events_engine';
+import { addNamespace, fireEvent } from '@js/events/utils/index';
+
+const DBLCLICK_EVENT_NAME = 'dxdblclick';
+const DBLCLICK_NAMESPACE = 'dxDblClick';
+const NAMESPACED_CLICK_EVENT = addNamespace(clickEventName, DBLCLICK_NAMESPACE);
+const DBLCLICK_TIMEOUT = 300;
+
+const DblClick = Class.inherit({
```
Comment (alexanderPolosatov) [inline packages/devextreme/js/__internal/events/dblclick_impl.ts]: Why we should have this inherit?

Comment (GoodDayForSurf) [reply packages/devextreme/js/__internal/events/dblclick_impl.ts]: i don't know, I just copied legacy code

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/mask.tests.js (hunk 3 above)
Comment (GoodDayForSurf) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/mask.tests.js]: i add back second click in this tests, it seems tests needs two clicks

File: packages/devextreme/testing/tests/DevExpress.ui.events/dblclick.tests.js (hunk 7 above)
Comment (GoodDayForSurf) [reply packages/devextreme/testing/tests/DevExpress.ui.events/dblclick.tests.js]: yep.. I renamed it to something more appropriate

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/mask.tests.js [hunk 9]
```diff
@@ -2101,10 +2101,11 @@ QUnit.module('format: caret boundaries', moduleConfig, () => {
 
         this.input.focus();
         this.clock.tick(CARET_TIMEOUT_DURATION);
+
         for(let i = 0; i < 2; ++i) {
             this.keyboard.caret(3);
             this.input.trigger('dxclick');
-            this.clock.tick(CARET_TIMEOUT_DURATION);
+            this.clock.tick(300);
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/mask.tests.js]: Let's move `300` to a constant `DBLCLICK_TIMEOUT` in the beginning of the file to explain this magic number

**Review by ksercs — DISMISSED** [review]
For part in NumberBox - approve.
I also created a backlog card to reconsider using of timeouts for masks.

**Review by dxvladislavvolkov — DISMISSED** [review]

**Review by dxvladislavvolkov — APPROVED** [review]

**Review by ksercs — APPROVED** [review]

**Review by vladaskorohodova — APPROVED** [review]

**Pull request merged**